031-statefulset-removal.md

```diff
@@ -0,0 +1,242 @@
+For example:
+* Every Pod would be named something like `my-cluster-kafka-1001-0` where `1001` would be an example of a broker ID and `0` is the suffix given to the Pod by the StatefulSet
+* We would need to manage multiple resources
+
+The main advantage of this approach is that Kubernetes would be still responsible for creating the Pods and we would not need to worry about it.
+
+### StatefulSets with mapping
+
+Another considered alternative was to keep using StatefulSets as today but use a special mapping to determine which StatefulSet index matches which broker ID.
+That might provide more flexibility since when the StatefulSet is being scaled down, the operator could just change the mapping to remove a broker with a specific ID regardless of which pod is being removed from the StatefulSet.
+However, while such mapping can be developed for the broker IDs, the StatefulSet does not control only the pod names but also the storage mounted into the pods.
+And while the mapping of the broker IDs can be easy changed, you cannot easily change the storage.
+So after the mapping change, we would end up with new broker IDs used by the pods, but with persistent volumes from the broker which used given pod previously.
+
+### Managing the Pods directly
+
+It would be also possible to manage the Pods from the `KafkaAssemblyOperator`.
+In the same way as it is currently creating the StatefulSets or deleting the Pods, it can also create and update the Pods directly.
+This would make things easier because we would not need any additional controller or custom resource.
+However, the Kafka CR reconciliation is currently a very _heavy_ process that does lot of things in a serial fashion.
+It manages both ZooKeeper and Kafka and it uses long-running reconciliation.
+The reconciliation often takes minutes while it waits for the Pods to be ready or for load balancers to be created.
+It runs multiple reconciliations in parallel, but for each Kafka custom resource only one reconciliation can be in progress.
+This means that it might often take a long time before the reconciliation gets to the point when it can recreate the Pods.
+Imagine a situation in which several Kafka brokers are unavailable because of a failure in one of the data centers, and the reconciliation is stuck waiting for a load balancer creation to finish while the pods are missing.
+A situation such as this can easily arise and would not be acceptable.
+
+Unlike the Pods, the StatefulSet is treated by Strimzi as a stable resource and used to carry different information such as:
+* Current storage configuration
+* Kafka versions
+* CA generations
+
+If we manage the Pods directly, we would also need to find a way how to replace them.
+
+## Compatibility
+
+The removal of StatefulSets is obviously a significant change.
+But the migration between StatefulSets and StrimziPodSets should be possible without any service interruption or availability issues.
```

Comment:
Will it be possible to seamlessly upgrade to, and downgrade from StrimziPodSets?

Reply:
Yes, it will be possible. Improved in the proposal.